central/role/datastore/datastore_impl.go

```diff
@@ -32,6 +32,117 @@ type dataStoreImpl struct {
 	lock sync.RWMutex
 }
 
+func (ds *dataStoreImpl) UpsertRole(ctx context.Context, newRole *storage.Role) error {
+	if err := sac.VerifyAuthzOK(roleSAC.WriteAllowed(ctx)); err != nil {
+		return err
+	}
+	if err := rolePkg.ValidateRole(newRole); err != nil {
+		return errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	oldRole, exists, err := ds.roleStorage.Get(ctx, newRole.GetName())
+	if err != nil {
+		return err
+	}
+	if exists {
+		if err := verifyRoleOriginMatches(ctx, oldRole); err != nil {
+			return err
+		}
+	}
+	if err := verifyRoleOriginMatches(ctx, newRole); err != nil {
+		return err
+	}
+
+	permissionSet, accessScope, err := ds.verifyRoleReferencesExist(ctx, newRole)
+	if err != nil {
+		return err
+	}
+	if err := verifyPermissionSetOriginMatches(ctx, permissionSet); err != nil {
+		return err
+	}
+	if err := verifyAccessScopeOriginMatches(ctx, accessScope); err != nil {
+		return err
+	}
```

Comment:
So in case someone would use the `UpsertRole` for this datastore, they could only do it in case the role doesn't reference a declaratively created access scope / permission set? Don't think we want to make this constraint just now.

Reply:
If I understand correctly, we want to allow imperative reference declarative, but not other way around.
Do you want to introduce this constraint within this PR, or move it to a separate ticket(this one, for example: https://issues.redhat.com/browse/ROX-14999)?

Reply:
I guess we can move it to the separate ticket, fine by me.